File: src/mcp_grafana/tools/incident.py
from datetime import datetime
from typing import Literal

from mcp.server import FastMCP
from pydantic import BaseModel, Field

from ..client import (
    grafana_client,
    AddActivityToIncidentArguments,
    CreateIncidentArguments,
)
from ..grafana_types import QueryIncidentPreviewsRequest, IncidentPreviewsQuery


class ListIncidentsArguments(BaseModel):
    """
    Arguments for the list_incidents tool.
    """

    # Note: this differs from what the client expects (which is a closer match
    # to the API) so that we can make it easier for LLM clients to use sensible
    # arguments without them having to know the query syntax.

    limit: int = Field(default=10, ge=1, le=100)
    drill: bool = Field(
        default=False, description="Whether to include drill incidents."
    )
    status: Literal["resolved", "active"] | None = Field(
        default=None,
        description="The status of the incidents to include. If not provided, all incidents will be included.",
    )


async def list_incidents(arguments: ListIncidentsArguments) -> bytes:
    """
    List incidents from the Grafana Incident incident management tool.

    Incidents will be returned in descending order by creation time.
    """
    query_string = "isdrill:false" if not arguments.drill else ""
    if arguments.status is not None:
        query_string += f" and status:{arguments.status}"
    body = QueryIncidentPreviewsRequest(
        query=IncidentPreviewsQuery(
            queryString=query_string,
            orderDirection="DESC",
            orderField="createdTime",
            limit=arguments.limit,
        ),
        includeCustomFieldValues=True,
    )
    return await grafana_client.list_incidents(body)


async def create_incident(arguments: CreateIncidentArguments) -> bytes:
    """
    Create an incident in the Grafana Incident incident management tool.
    """
    return await grafana_client.create_incident(arguments)


async def add_activity_to_incident(
    incident_id: str, body: str, event_time: datetime | None = None
) -> bytes:
    """
    Add an activity to an incident in the Grafana Incident incident management tool.

    :param incident_id: The ID of the incident to add the activity to.
    :param body: The body of the activity. URLs will be parsed and attached as context.
    :param event_time: The time that the activity occurred. If not provided, the current time will be used.
                       If provided, it must be in RFC3339 format.
    """
    return await grafana_client.add_activity_to_incident(
        AddActivityToIncidentArguments(
            incidentId=incident_id,
            body=body,
            activityKind="userNote",
            eventTime=event_time,
        )
    )


async def resolve_incident(incident_id: str, summary: str) -> bytes:
    """
    Resolve an incident in the Grafana Incident incident management tool.

    :param incident_id: The ID of the incident to resolve.
    :param summary: The summary of the incident now it has been resolved.
                    This should be succint but thorough and informative,
                    enough to serve as a mini post-incident-report.
    """
    return await grafana_client.close_incident(incident_id, summary)


def add_tools(mcp: FastMCP):
    mcp.add_tool(list_incidents)
    mcp.add_tool(create_incident)
    mcp.add_tool(add_activity_to_incident)
    mcp.add_tool(resolve_incident)

File: tools/incident.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/grafana/incident-go"
	mcpgrafana "github.com/grafana/mcp-grafana"
	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

type ListIncidentsParams struct {
	Limit  int    `json:"limit" jsonschema:"description=The maximum number of incidents to return"`
	Drill  bool   `json:"drill" jsonschema:"description=Whether to include drill incidents"`
	Status string `json:"status" jsonschema:"description=The status of the incidents to include"`
}

func listIncidents(ctx context.Context, args ListIncidentsParams) (*incident.QueryIncidentsResponse, error) {
	c := mcpgrafana.IncidentClientFromContext(ctx)
	is := incident.NewIncidentsService(c)
	query := ""
	if !args.Drill {
		query = "isdrill:false"
	}
	if args.Status != "" {
		query += fmt.Sprintf(" and status:%s", args.Status)
	}
	incidents, err := is.QueryIncidents(ctx, incident.QueryIncidentsRequest{
		Query: incident.IncidentsQuery{
			QueryString:    query,
			OrderDirection: "DESC",
			Limit:          args.Limit,
		},
	})
	if err != nil {
		return nil, fmt.Errorf("list incidents: %w", err)
	}
	return incidents, nil
}

var ListIncidents = mcpgrafana.MustTool(
	"list_incidents",
	"List incidents",
	listIncidents,
)

type CreateIncidentParams struct {
	Title         string                   `json:"title" jsonschema:"description=The title of the incident"`
	Severity      string                   `json:"severity" jsonschema:"description=The severity of the incident"`
	RoomPrefix    string                   `json:"roomPrefix" jsonschema:"description=The prefix of the room to create the incident in"`
	IsDrill       bool                     `json:"isDrill" jsonschema:"description=Whether the incident is a drill incident"`
	Status        string                   `json:"status" jsonschema:"description=The status of the incident"`
	AttachCaption string                   `json:"attachCaption" jsonschema:"description=The caption of the attachment"`
	AttachURL     string                   `json:"attachUrl" jsonschema:"description=The URL of the attachment"`
	Labels        []incident.IncidentLabel `json:"labels" jsonschema:"description=The labels to add to the incident"`
}

func createIncident(ctx context.Context, args CreateIncidentParams) (*mcp.CallToolResult, error) {
	c := mcpgrafana.IncidentClientFromContext(ctx)
	is := incident.NewIncidentsService(c)
	incident, err := is.CreateIncident(ctx, incident.CreateIncidentRequest{
		Title:         args.Title,
		Severity:      args.Severity,
		RoomPrefix:    args.RoomPrefix,
		IsDrill:       args.IsDrill,
		Status:        args.Status,
		AttachCaption: args.AttachCaption,
		AttachURL:     args.AttachURL,
		Labels:        args.Labels,
	})
	if err != nil {
		return nil, fmt.Errorf("create incident: %w", err)
	}
	b, err := json.Marshal(incident)
	if err != nil {
		return nil, fmt.Errorf("marshal incident: %w", err)
	}
	return mcp.NewToolResultText(string(b)), nil
}

var CreateIncident = mcpgrafana.MustTool(
	"create_incident",
	"Create an incident",
	createIncident,
)

type AddActivityToIncidentParams struct {
	IncidentID string `json:"incidentId" jsonschema:"description=The ID of the incident to add the activity to"`
	Body       string `json:"body" jsonschema:"description=The body of the activity. URLs will be parsed and attached as context"`
	EventTime  string `json:"eventTime" jsonschema:"description=The time that the activity occurred. If not provided, the current time will be used"`
}

func addActivityToIncident(ctx context.Context, args AddActivityToIncidentParams) (*mcp.CallToolResult, error) {
	c := mcpgrafana.IncidentClientFromContext(ctx)
	as := incident.NewActivityService(c)
	activity, err := as.AddActivity(ctx, incident.AddActivityRequest{
		IncidentID:   args.IncidentID,
		ActivityKind: "userNote",
		Body:         args.Body,
		EventTime:    args.EventTime,
	})
	if err != nil {
		return nil, fmt.Errorf("add activity to incident: %w", err)
	}
	b, err := json.Marshal(activity)
	if err != nil {
		return nil, fmt.Errorf("marshal incident: %w", err)
	}
	return mcp.NewToolResultText(string(b)), nil
}

var AddActivityToIncident = mcpgrafana.MustTool(
	"add_activity_to_incident",
	"Add an activity to an incident",
	addActivityToIncident,
)

func AddIncidentTools(mcp *server.MCPServer) {
	ListIncidents.Register(mcp)
	CreateIncident.Register(mcp)
	AddActivityToIncident.Register(mcp)
}
